pkg/cmd/experimental/syncgroups/groupsyncer.go

```diff
@@ -0,0 +1,260 @@
+	// Allows the Syncer to search for OpenShift Groups
+	groupClient client.GroupInterface
+	// config stores all sync configuration
+	config syncconfig.LDAPSyncConfig
+}
+
+// Sync allows the LDAPGroupSyncer to be a GroupSyncer
+func (s *LDAPGroupSyncer) Sync() error {
+	// determine what to sync
+	ldapGroupUIDs, err := s.groupLister.ListGroups()
+	if err != nil {
+		return err
+	}
+
+	for _, ldapGroupUID := range ldapGroupUIDs {
+		// get membership data
+		memberEntries, err := s.groupMemberExtractor.ExtractMembers(ldapGroupUID)
+		if err != nil {
+			return err
+		}
+
+		// determine OpenShift Users' usernames for LDAP group members
+		usernames, err := s.determineUsernames(memberEntries)
+		if err != nil {
+			return err
+		}
+
+		// find OpenShift Group to update with this record
+		group, err := s.findOrCreateGroup(ldapGroupUID)
+		if err != nil {
+			return err
+		}
+
+		err = s.updateGroup(group, usernames)
+		if err != nil {
+			return err
+		}
+
+	}
+	return nil
+}
+
+// determineUsers determines the OpenShift Users that correspond to a list of LDAP member entries
+func (s *LDAPGroupSyncer) determineUsernames(members []*ldap.Entry) ([]string, error) {
+	var usernames []string
+	for _, member := range members {
+		username, err := s.userNameMapper.UserNameFor(member)
+		if err != nil {
+			return nil, err
+		}
+		usernames = append(usernames, username)
+	}
+	return usernames, nil
+}
+
+func (s *LDAPGroupSyncer) findOrCreateGroup(ldapGroupUID string) (*ouserapi.Group, error) {
+	groupName, err := s.groupNameMapper.GroupNameFor(ldapGroupUID)
+	if err != nil {
+		return nil, err
+	}
+
+	group, err := s.groupClient.Get(groupName)
+
+	if err != nil && !kerrs.IsNotFound(err) {
+		// Fail on errors other than "not found"
+		return nil, err
+	}
+
+	if err != nil && kerrs.IsNotFound(err) {
+		if s.config.ProvisionGroups {
+			// Try to create a group with the available name
+			desiredGroup := &ouserapi.Group{
+				ObjectMeta: kapi.ObjectMeta{
+					Name: groupName,
+				},
+			}
+
+			// Create the group
+			createdGroup, err := s.groupClient.Create(desiredGroup)
+			if err != nil {
+				return nil, err
+			}
+
+			s.addAnnotations(createdGroup, ldapGroupUID)
+			return createdGroup, nil
+		}
+		return nil, fmt.Errorf("no group found with name: %s", groupName)
+	}
+	s.addAnnotations(group, ldapGroupUID)
+	return group, nil
+}
+
+func (s *LDAPGroupSyncer) addAnnotations(group *ouserapi.Group, ldapGroupUID string) {
+	group.Annotations[syncconfig.LDAPURL] = s.config.ServerConfig.Host
+	group.Annotations[syncconfig.LDAPUID] = ldapGroupUID
+	group.Annotations[syncconfig.LDAPSyncTime] = strconv.FormatInt(time.Now().Unix(), 10)
+}
+
+func (s *LDAPGroupSyncer) updateGroup(group *ouserapi.Group, usernames []string) error {
+	// overwrite Group Users data
+	group.Users = usernames
+
+	_, err := s.groupClient.Update(group)
+	return err
+}
+
+func New(config syncconfig.LDAPSyncConfig) (*LDAPGroupSyncer, error) {
```

Comment:
No, this is insane to read.  You need to decompose how the pieces are built.  This `New` method should contain the interfaces used by `LDAPGroupSyncer`.  Those individual interfaces should be built with much tighter argument lists.    In fact, I'd say you should probably not have a single exposed method that accepts enough information to build more than one kind of `GroupLister` as a for instance.